python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

```diff
@@ -183,6 +191,27 @@ def _prepare_llm_request(self, row: Dict[str, Any]) -> vLLMEngineRequest:
         else:
             image = []
 
+        # The request is for LoRA.
+        lora_request = None
+        if "model" in row and row["model"] != self.model:
+            if self.vllm_use_v1:
+                raise ValueError("We do not support LoRA with vLLM v1")
+
+            lora_name = row["model"]
+            if lora_name not in self.lora_name_to_request:
+                # Load a new LoRA adapter if it is not loaded yet.
+                with self.lora_lock:
+                    # Make sure no other thread has loaded the same LoRA adapter.
+                    if lora_name not in self.lora_name_to_request:
```

Comment:
Is this check again necessary?

Reply:
I think so? If you have multiple rows getting to the lock at the same time. When one of them gets in and load the LoRA adapter, and the rest should just bypass.

Reply:
I think a better way here is to create a thread-safe Cache Object for lora_name_to_request. It would be sth like:
```
class Cache:
    def __init__(self):
        self._cache = {}
        self._lock = asyncio.Lock()
    
    async def get(self, key):
        async with self._lock:
            return self._cache.get(key)
    
    async def set(self, key, value):
        async with self._lock:
            self._cache[key] = value
```

Reply:
After the second thought I feel we don't need a separate object for this:
1. Even with this object we still need to check `key in self._cache` in `.set()` when a task acquires the lock.
2. The LoRA ID is accumulated outside of this object, so once we generate `value`, we have to set it; otherwise some LoRA IDs will be missed (although this is not a very serious issue given that LoRA ID is more like an identifier).
3. We currently don't have other use cases, so probably no need to over engineering atm.

Reply:
👍